src/transaction.rs

```diff
@@ -104,46 +102,37 @@ where
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         let delivery_result = match self.work.poll() {
             Ok(Async::NotReady) => return Ok(Async::NotReady),
-            Err(err) => {
+            Err(deadline_error) => {
                 let duration = self.start_time.elapsed();
-                // Request errored
-                match err {
-                    Either::A((hyper_error, _timeout)) => {
-                        trace!("Transaction errored during delivery, error: {:?}, duration: {:?}", hyper_error, duration);
-                        DeliveryResult::HyperError {
-                            error: hyper_error,
-                            duration,
-                        }
-                    },
-                    // Timeout errored
-                    Either::B((timeout_error, _request)) => {
-                        trace!("Transaction errored during timeout, error: {}, duration: {:?}", timeout_error, duration);
-                        DeliveryResult::TimeoutError {
-                            error: timeout_error,
-                            duration,
-                        }
+                if deadline_error.is_timer() {
+                    let timer_error = deadline_error.into_timer().expect("is_timer -> into_timer");
+                    trace!("Timer around Transaction errored, error: {:?}, duration: {:?}", timer_error, duration);
+                    DeliveryResult::TimeoutError {
+                        error: io::Error::new(ErrorKind::Other, timer_error),
+                        duration,
+                    }
+                } else if deadline_error.is_elapsed() {
+                    DeliveryResult::Timeout {
+                        duration,
+                    }
+                } else if deadline_error.is_inner() {
+                    let hyper_error = deadline_error.into_inner().expect("is_inner -> into_inner");
+                    trace!("Transaction errored during delivery, error: {:?}, duration: {:?}", hyper_error, duration);
+                    DeliveryResult::HyperError {
+                        error: hyper_error,
+                        duration,
                     }
+                } else {
+                    unreachable!();
```

Comment:
If a future update to Tokio adds another kind of error to this `DeadlineError` type and this `unreachable!` statement gets hit in production, there will be no useful backtrace to tell us where it happened. It would be good to add some kind of easily greppable message into this `unreachable!` statement.

Reply:
👍